src/core/qgspointlocator.h

```diff
@@ -287,11 +287,20 @@ class CORE_EXPORT QgsPointLocator : public QObject
 
             if ( !( geom.isNull() || geom.isEmpty() ) )
             {
-              const QgsLineString line( geom.vertexAt( mVertexIndex ), geom.vertexAt( mVertexIndex + 1 ) );
-              point = QgsGeometryUtils::closestPoint( line, QgsPoint( snappedPoint ) );
+              // when snapping to a curve we need to use real geometry in order to have correct location
+              // of the snapped point, see https://github.com/qgis/QGIS/issues/53197.
+              // In other cases it is ok to use only a segment to speedup calculations.
+              if ( QgsWkbTypes::isCurvedType( mLayer->wkbType() ) )
```

Comment:
Could we not use `geom.wkbType()` here so the optimized route is used for non-curved geometries within curved layers?

Reply:
Changed to use `geom.wkbType()`